### PR TITLE
CDK: Emit control message on config mutation

### LINKED ISSUE #17910: Python CDK Updated to use and emit updated configurations
_Part of the [Single Use Refresh Tokens](https://github.com/airbytehq/airbyte/issues/3990) Epic_

Add tooling to the Python CDK so that a connector can update it's in-memory configuration object, and when it does, it will emit a `AirbyteControlMessage.ConnectorConfig` message to share that new configuration with the platform so it can be stored.

1. All oAuth connectors which use the CDK will emit a clear and actional error message to users when the oAuth flow fails due to token/access errors.  AirbyteTraceMessage.Error.external_message will be used to inform the user that they need to restart the oAuth workflow from the source’s config page. 
2. For the specific use case of persisting config updates for access/refresh token we can take the following approach: AbstractOauth2Authenticator takes an emit_config_update boolean attribute, the get_access_token and refresh_access_token methods emits AirbyteControlMessage.ConnectorConfig if emit_config_update == true
3. For a more global approach for sending config updates: config are currently plain python dictionary, we could:
    a. Load this config dictionary into a Config class with mutable attributes.
    b. Calls on setter of the Config object could emit AirbyteControlMessage.ConnectorConfig

As part of this work, a connector should be updated to show that single-use refresh tokens work.  We think that Jira is a good connector to test with.

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Closes #17910
Quoting @davinchia in #3990:
> Today we do not handle the case where a new refresh token is returned when a new access token is granted.

A [new protocol message](https://github.com/airbytehq/airbyte/pull/17907):  `AirbyteControlMessage.ConnectorConfig` was introduced to enable the connector to share a configuration update with the platform.

This PR is an attempt to update the CDK to enable automatic emission of `AirbyteControlMessage.ConnectorConfig` on mutation of a `config` `dict` keys, and nested keys.
 
## How
1. Add a `config` kwarg to `OAuth2Authenticator.__init__`. Observe this `config` by creating an `ObservedDict` assigned to `self.connector_config`
2. `ObservedDict` get created with a `ConfigObserver` object. `ObservedDict` call `ConfigObserver.update` on `__set_item__` calls (this is the function called on key assignment: e.g `config["refresh_token"] = "new_refresh_token"`.
3. `ConfigObserver.update`  emits an `AirbyteControlMessage.ConnectorConfig` with the updated config.

## N.B.
I chose the observer pattern because I thought this could be a silver-bullet-like approach to enable additional future use cases around `AirbyteControlMessage.ConnectorConfig` message. 

**The platform is not currently processing these new messages, so emitting  will `AirbyteControlMessage.ConnectorConfig` will currently have no effect**/

## Recommended reading order
1. https://github.com/airbytehq/airbyte/blob/12db1b7b0de9c5fc6a7d9ae5cba2e382332617d7/airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py#L31
2. https://github.com/airbytehq/airbyte/blob/12db1b7b0de9c5fc6a7d9ae5cba2e382332617d7/airbyte-cdk/python/airbyte_cdk/config_observation.py#L14
3. https://github.com/airbytehq/airbyte/blob/12db1b7b0de9c5fc6a7d9ae5cba2e382332617d7/airbyte-cdk/python/unit_tests/sources/streams/http/requests_native_auth/test_requests_native_auth.py#L178
4. https://github.com/airbytehq/airbyte/blob/12db1b7b0de9c5fc6a7d9ae5cba2e382332617d7/airbyte-cdk/python/unit_tests/test_config_observation.py#L12

## 🚨 User Impact 🚨
To benefit from config observation, the connector developer can implement a custom authenticator inheriting from OAuth2Authenticator and create any token-refresh logic in this authenticator. The functions implementing this logic will have to mutate `self.connector_config` to emit a control message.

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/utils/schema_helpers.py [hunk 1]
```diff
@@ -176,23 +176,18 @@ def dict(self, *args, **kwargs):
         return super().dict(*args, **kwargs)
 
 
-def split_config(config: Mapping[str, Any]) -> Tuple[dict, InternalConfig]:
+def split_config(config: MutableMapping[str, Any]) -> Tuple[MutableMapping, InternalConfig]:
```
Comment (erohmensing) [inline airbyte-cdk/python/airbyte_cdk/sources/utils/schema_helpers.py:179]: Here the use case of popping items from the original config makes sense if the object passed to it is indeed always loaded from the config file, but since we can't validate that from within the function, I wonder if this could lead to unexpected usage elsewhere re: changing the original dict. Maybe renaming to something like `filter_internal_keywords` would make it clearer both what happens in the mutation (i.e. if you were to call it without using the return values) as well as what the return values are

Comment (erohmensing) [reply airbyte-cdk/python/airbyte_cdk/sources/utils/schema_helpers.py:179]: My only comment -- otherwise I took a look, really cool stuff 👀

**Review by sherifnada — COMMENTED** [review]
@alafanechere could you show an example of how would this be used in the context of an HTTPStream and a single-use oauth authenticator?

File: airbyte-cdk/python/airbyte_cdk/sources/utils/schema_helpers.py (hunk 1 above)
Comment (alafanechere) [reply airbyte-cdk/python/airbyte_cdk/sources/utils/schema_helpers.py]: 👍 Yes you're right! I renamed the function to `filter_internal_keywords`

File: airbyte-cdk/python/unit_tests/test_config_observation.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        mock_observer = mocker.Mock()
+        my_observed_dict = ObservedDict({"key": "value"}, mock_observer)
+        assert mock_observer.update.call_count == 0
+        my_observed_dict["key"] = {"nested_key": "nested_value"}
+        assert mock_observer.update.call_count == 1
+        my_observed_dict["key"]["nested_key"] = "new_nested_value"
+        assert mock_observer.update.call_count == 2
+        # Setting the same value again should not call observer's update
+        my_observed_dict["key"]["nested_key"] = "new_nested_value"
+        assert mock_observer.update.call_count == 2
```
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-cdk/python/unit_tests/test_config_observation.py]: I changed the default logic to: always emit the message, even if the re-assigned value is the same.

**Comment by alafanechere** [comment]
@sherifnada I achieved a working example for BingAds in https://github.com/airbytehq/airbyte/pull/19428/commits/733b89a4bd9de3b6823a7d318203c94340bff51b (will revert afterward) 

A `READ` outputs the following to stdout (I manually removed sensitive fields):
```bash
$ python main.py read --catalog integration_tests/configured_catalog.json --config secrets/config.json
{"type": "LOG", "log": {"level": "INFO", "message": "Starting syncing SourceBingAds"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Fetching access token ..."}}
{"type": "CONNECTOR_CONFIG", "emitted_at": 1668613492681.509, "connectorConfig": {"config": {"tenant_id": "common", "refresh_token": "<OBFUSCATED-FOR-DEMO>", "client_id": "<OBFUSCATED-FOR-DEMO>", "developer_token": "<OBFUSCATED-FOR-DEMO>", "reports_start_date": "2020-02-25", "hourly_reports": false, "daily_reports": false, "weekly_reports": true, "monthly_reports": true}}}
{"type": "LOG", "log": {"level": "INFO", "message": "Syncing stream: ad_groups "}}
... Records are coming in afterward
```

**Comment by evantahler** [comment]
```
{"type": "CONNECTOR_CONFIG", "emitted_at": 1668613492681.509, "connectorConfig": {...}}
```

I think there's one more envelope layer on that message, and it should be something like:

```
{type: "CONTROL", control: {"type": "CONNECTOR_CONFIG", "emitted_at": 1668613492681.509, "connectorConfig": {...}}}
```

**Comment by alafanechere** [comment]
> ```
> {"type": "CONNECTOR_CONFIG", "emitted_at": 123, "connectorConfig": {...}}
> ```
> 
> I think there's one more envelope layer on that message, and it should be something like:
> 
> ```
> {"type": "CONTROL", "control": {"type": "CONNECTOR_CONFIG", "emitted_at": 123, "connectorConfig": {...}}}
> ```

I think I forgot to wrap this around an AirbyteMessage right?

**Comment by evantahler** [comment]
> I think I forgot to wrap this around an AirbyteMessage right?
That's probably it!

**Comment by alafanechere** [comment]
```bash
$ python main.py read --catalog integration_tests/configured_catalog.json --config secrets/config.json
{"type": "LOG", "log": {"level": "INFO", "message": "Starting syncing SourceBingAds"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Fetching access token ..."}}
{"type": "CONTROL", "log": null, "spec": null, "connectionStatus": null, "catalog": null, "record": null, "state": null, "trace": null, "control": {"type": "CONNECTOR_CONFIG", "emitted_at": 1668615971298.429, "connectorConfig": {"config": {"tenant_id": "common", "refresh_token": "<MANUALLY_OBFUSCATED>", "client_id": "<MANUALLY_OBFUSCATED>", "developer_token": "<MANUALLY_OBFUSCATED>", "reports_start_date": "2020-02-25", "hourly_reports": false, "daily_reports": false, "weekly_reports": true, "monthly_reports": true}}}}
{"type": "LOG", "log": {"level": "INFO", "message": "Syncing stream: ad_groups "}}
... Records are coming in afterward
```
@evantahler better now? 😄

**Comment by evantahler** [comment]
Those nulls seem kind of unnecessary, but yes!

**Review by sherifnada — CHANGES_REQUESTED** [review]
I am worried the current approach could result in surprises for the developer, and maybe even in a way that breaks the connector's configuration. There's a number of connectors already which put stuff in the `config` which are nonsensical to processes outside the connector for convenience (e.g: some connectors put python objects in the `config` dict just to help pass stuff around). In some cases, they even overwrite existing keys with those values. 

Is there a way to scope emitting CONTROL messages to the exact use case we are trying to solve for (single-use refresh tokens)? 

The outlines of the solution I would suggest is to add something to the oauth authenticator which indicates that this is a single use refresh token API, and that everytime it's updated, the authenticator should output a CONTROL message. 

Maybe something like: 

```
OauthAuthenticator( <args>...., single_use_refresh_token=True, connector_config=config)
```

where the connector knows that if single_use_refresh_token is true, then it must also be provided a config, and to emit a CONTROL message with the new refresh token whenever it receives one.

**Comment by alafanechere** [comment]
Thanks for raising these concerns @sherifnada . 
I agree that I'd prefer not to introduce surprising side effects, so thanks for sharing that mutating config is an existing behavior. 
I kept the original work on the creation of `ObservedDict` / `ConfigObserver` objects, but delegated their usual creation to the `OAuth2Authenticator`. 
If `config` is passed to `OAuth2Authenticator`, it's copied to a `self.connector_config` attribute, which is observed. If mutation of the `self.connector_config` happens in the authenticator context a control message will be emitted. 
To benefit from this, the connector developer can implement a custom authenticator inheriting from `OAuth2Authenticator` and create any token-refresh logic in this authenticator. The functions implementing this logic will have to mutate `self.connector_config` to emit a control message.

The committed work on the Bing Ads connector does not represent the interaction with `OAuth2Authenticator` as this connector uses a custom client class. But my changes on this connector show that it's pretty easy to benefit from config observation in another context.

Let me know if an existing OAuth connector could benefit from this new feature and I'll try to implement it.

File: airbyte-cdk/python/airbyte_cdk/config_observation.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    def update(self):
+        ...
+
+
+class ObservedDict(dict):
+    def __init__(self, non_observed_mapping: MutableMapping, observer: BaseObserver, update_on_unchanged_value=True) -> None:
+        self.observer = observer
+        self.update_on_unchanged_value = update_on_unchanged_value
+        for item, value in self.items():
+            if isinstance(value, MutableMapping):
```
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-cdk/python/airbyte_cdk/config_observation.py:20]: I tried to handle list of dicts in the if statement below.

File: airbyte-cdk/python/airbyte_cdk/sources/utils/schema_helpers.py [hunk 4]
```diff
@@ -176,23 +176,18 @@ def dict(self, *args, **kwargs):
...
     internal_config = {}
-    for k, v in config.items():
-        if k in InternalConfig.KEYWORDS:
-            internal_config[k] = v
-        else:
-            main_config[k] = v
-    return main_config, InternalConfig.parse_obj(internal_config)
+    for key in list(config.keys()):
+        if key in InternalConfig.KEYWORDS:
+            internal_config[key] = config.pop(key)
```
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-cdk/python/airbyte_cdk/sources/utils/schema_helpers.py]: I reverted this change as config observation only happens in the authenticator context now.

File: airbyte-cdk/python/airbyte_cdk/config_observation.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+from airbyte_cdk.models import AirbyteControlConnectorConfigMessage, AirbyteControlMessage, AirbyteMessage, OrchestratorType, Type
+
+
+class ObservedDict(dict):
+    def __init__(self, non_observed_mapping: MutableMapping, observer: ConfigObserver, update_on_unchanged_value=True) -> None:
+        self.observer = observer
+        self.update_on_unchanged_value = update_on_unchanged_value
+        for item, value in non_observed_mapping.items():
+            # Observe nested dicts
+            if isinstance(value, MutableMapping):
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/config_observation.py:20]: Can this be DRYd with the logic in set item?

File: airbyte-cdk/python/airbyte_cdk/config_observation.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                if isinstance(sub_value, MutableMapping):
+                    value[i] = ObservedDict(sub_value, self.observer)
+        super(ObservedDict, self).__setitem__(item, value)
+        if self.update_on_unchanged_value or value != previous_value:
+            self.observer.update()
+
+
+class ConfigObserver:
+    """This class is made to track mutations on ObservedDict config.
+    When update is called the observed configuration is saved on disk a CONNECTOR_CONFIG control message is emitted on stdout.
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/config_observation.py:49]: Is this true? I'm not seeing a write to disk

File: airbyte-cdk/python/airbyte_cdk/config_observation.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            emitted_at=time.time() * 1000,
+            connectorConfig=AirbyteControlConnectorConfigMessage(config=self.config),
+        )
+        airbyte_message = AirbyteMessage(type=Type.CONTROL, control=control_message)
+        print(airbyte_message.json(exclude_unset=True))
+
+
+def observe_connector_config(non_observed_connector_config: MutableMapping[str, Any]):
+    if isinstance(non_observed_connector_config, ObservedDict):
+        raise ValueError("This connector configuration is already observed")
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/config_observation.py:70]: Any reason not to make this re-entrant?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py [hunk 8]
```diff
@@ -39,6 +42,7 @@ def __init__(
 
         self._token_expiry_date = token_expiry_date or pendulum.now().subtract(days=1)
         self._access_token = None
+        self._connector_config = observe_connector_config(connector_config) if connector_config else None
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py:45]: I think this won't happen automatically, we need the Authenticator to also update this particular config

**Review by sherifnada — COMMENTED** [review]
@alafanechere apologies for the late response and thank you for making these changes! 

One more request: I think this should be behavior in an off the shelf Authenticator provided with the CDK. IMO The abstraction that makes sense for the developer is: they need to be able to indicate the refresh token is single use, give us the full config, and tell us how to find the new refresh token from the API response. My guess is that we can look horizontally at the 7-8 APIs we know of which implement this functionality and create an abstraction that works across all of them.

File: airbyte-cdk/python/airbyte_cdk/config_observation.py (hunk 6 above)
Comment (alafanechere) [reply airbyte-cdk/python/airbyte_cdk/config_observation.py:49]: Oh yes, I removed the write to disk logic but forgot to update the docstring.

File: airbyte-cdk/python/airbyte_cdk/config_observation.py (hunk 7 above)
Comment (alafanechere) [reply airbyte-cdk/python/airbyte_cdk/config_observation.py:70]: I'd ask the opposite :) To limit side effects I'd prefer to ensure a root config that is already observed not to be re-observed.

**Comment by alafanechere** [comment]
> they need to be able to indicate the refresh token is single use, give us the full config, and tell us how to find the new refresh token from the API response

I would love to do this. But they would also need to tell us where to find the refresh_token key in the config so we can mutate it on their behalf. I'll try to do this.

**Comment by alafanechere** [comment]
@sherifnada I tried to go in the direction you suggested by implementing a specific `SingleUseRefreshTokenOauth2Authenticator` authenticator with a behavior that can match most existing OAuth scenarios. Let me know if it's what you had in mind 😄

**Comment by alafanechere** [comment]
I tried to use `source-harvest` as an example: https://github.com/airbytehq/airbyte/pull/19428/commits/c5a0b6d010699781668094798b9ee42c3009eb55

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py [hunk 9]
```diff
@@ -84,3 +82,109 @@ def access_token(self) -> str:
...
+        )
+
+    def _validate_connector_config(self):
+        """Validates the defined getters for configuration values are returning values.
+
+        Raises:
+            ValueError: Raised if the defined getters are not returning a value.
+        """
+        for field_name, getter in [
+            ("client_id", self.get_client_id),
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py:133]: why do we care about retaining and validating client_id and client_secret in this class instead of just passing up the concern to the super?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py [hunk 10]
```diff
@@ -84,3 +82,109 @@ def access_token(self) -> str:
...
+            except (AssertionError, KeyError):
+                raise ValueError(
+                    f"This authenticator expects a {field_name} field under the {self.credentials_configuration_field_name} field. Please override this class getters or change your configuration structure."
+                )
+
+    def get_refresh_token_name(self) -> str:
+        return self._refresh_token_name
+
+    def _get_config_credentials_field(self, field_name):
+        return self._connector_config[self.credentials_configuration_field_name][field_name]
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py:148]: we should use `dpath` to allow easily removing any assumptions about where in the code the refresh/access tokens live

File: airbyte-integrations/connectors/source-harvest/source_harvest/source.py [hunk 11]
```diff
@@ -54,11 +54,8 @@ def get_authenticator(config):
         credentials = config.get("credentials", {})
         if credentials and "client_id" in credentials:
             return HarvestOauth2Authenticator(
+                config,
                 token_refresh_endpoint="https://id.getharvest.com/api/v2/oauth2/token",
-                client_id=credentials.get("client_id"),
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-harvest/source_harvest/source.py:58]: why were these removed?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py [hunk 12]
```diff
@@ -80,3 +82,109 @@ def access_token(self) -> str:
...
+            self.access_token = new_access_token
+            self.set_token_expiry_date(t0.add(seconds=access_token_expires_in))
+            self.set_refresh_token(new_refresh_token)
+        return self.access_token
+
+    def refresh_access_token(self) -> Tuple[str, int, str]:
+        try:
+            response_json = self._get_refresh_access_token_response()
+            return (
+                response_json[self.get_access_token_name()],
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py:185]: would recommend using `dpath`

**Review by sherifnada — COMMENTED** [review]
nice! looking really good. Some questions

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py (hunk 9 above)
Comment (alafanechere) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py:133]: As `client_id`, `client_secret`, and `refresh_token` are mandatory arguments of the parent class, and the current constructor signature of the new class only takes a `connector_config` dict, I thought it would be safer to fail early if these field value could not be parsed from the config.

File: airbyte-integrations/connectors/source-harvest/source_harvest/source.py (hunk 11 above)
Comment (alafanechere) [reply airbyte-integrations/connectors/source-harvest/source_harvest/source.py:58]: The `client_id` parameter does not exist in my `SingleUseRefreshTokenOauth2Authenticator` implementation. The `client_id` is retrieved from the configuration with dpath

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py (hunk 12 above)
Comment (alafanechere) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py:206]: Agreed but for implementation consistency with parent class and abstract class, I'd prefer to do this in a separate PR in which we can replace the `access_token_name`, `expires_in_name`, and `refresh_token_name` by dpaths that can be used when parsing these responses. Wdyt?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py (hunk 10 above)
Comment (alafanechere) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py:169]: Thanks, that was definitely what I was missing, I was not aware of the existence of this lib 😄

**Comment by alafanechere** [comment]
@sherifnada I used `dpath` to declare path to the config field that we use to build refresh requests and set a new refresh token values in the config. I think this is far better now for DX 😄

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py (hunk 12 above)
Comment (sherifnada) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py:206]: makes sens!

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py [hunk 13]
```diff
@@ -104,9 +105,29 @@ def __init__(
         refresh_token_name: str = "refresh_token",
         refresh_request_body: Mapping[str, Any] = None,
         grant_type: str = "refresh_token",
-        credentials_configuration_field_name: str = "credentials",
+        client_id_config_path="/credentials/client_id",
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py:108]: I would recommend to make this a list instead to avoid problems with special characters (dpath supports it like `["credentials", "client_secret"]`

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py [hunk 14]
```diff
@@ -80,3 +83,129 @@ def access_token(self) -> str:
...
+                assert getter()
+            except KeyError:
+                raise ValueError(
+                    f"This authenticator expects a value under the {field_path} field path. Please check your configuration structure or change the {parameter_name} value at initialization of this authenticator."
+                )
+
+    def get_refresh_token_name(self) -> str:
+        return self._refresh_token_name
+
+    def _get_config_credentials_field(self, field_name):
```
Comment (sherifnada) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py:168]: i think we can remove this method now, it seems unused

**Review by sherifnada — APPROVED** [review]
Good stuff! requesting a couple small changes but no extra review needed from my side. Thank you for making this awesome change.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py (hunk 13 above)
Comment (alafanechere) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py]: 👍 I defaulted to tuples and declared these parameters as `Sequence` type that dpath.get can handle according to what I understood from their implementation.
